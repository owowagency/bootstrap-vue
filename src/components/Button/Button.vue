<template>
    <Component
        :is="is"
        class="btn"
        :class="classes"
        :disabled="disabled"
        :href="href"
        :to="to"
    >
        <!-- @slot Displays content inside the button -->
        <slot />
    </Component>
</template>

<script lang="ts">
import {PropType, computed} from 'vue';
import useDisabled, {disabledProps} from '@/composables/useDisabled';
import useSize, {sizeProps} from '@/composables/useSize';
import useVariant, {variantProps} from '@/composables/useVariant';
import useClasses from '@/composables/useClasses';
</script>

<script lang="ts" setup>
const props = defineProps({
    ...disabledProps,
    href: {
        type: String,
        default: undefined,
    },
    is: {
        type: String,
        default: undefined,
    },
    /**
     * Changes the button to an outline variant
     */
    outline: {
        type: Boolean,
        default: false,
    },
    ...sizeProps,
    to: {
        type: [Object, String] as PropType<string|Record<string, unknown>>,
        default: undefined,
    },
    ...variantProps,
});

const variantClassTemplate = computed(() => `btn-${props.outline ? 'outline-' : ''}{0}`);

const {classes} = useClasses(computed(() => [
    useDisabled(props.disabled).disabledClass.value,
    useSize(props.size, 'btn-{0}').sizeClass.value,
    useVariant(props.variant, variantClassTemplate).variantClass.value,
]));

const is = computed(() => {
    if (props.is) {
        return props.is;
    }

    if (props.to) {
        return 'RouterLink';
    }

    if (props.href) {
        return 'a';
    }

    return 'button';
});
</script>

<docs>
```vue
<Button variant="danger">
    Delete
</Button>
```

## Imported Props

| Prop name | Description                  | Type                                             | Values | Default   |
| --------- | ---------------------------- | ------------------------------------------------ | ------ | --------- |
| disabled  | Disables the component       | boolean                                          | -      | false     |
| size      | The size of the component    | [Size](../../composables/useSize) (string)       | -      | 'md'      |
| variant   | The variant of the component | [Variant](../../composables/useVariant) (string) | -      | 'primary' |
</docs>
