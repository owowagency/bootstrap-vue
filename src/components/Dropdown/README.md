# Dropdown

```vue
<Dropdown :items="[Item]" />
```

## Imported Props

| Prop name | Description                                    | Type                                             | Values | Default   |
| --------- | ---------------------------------------------- | ------------------------------------------------ | ------ | --------- |
| items     | The items of the dropdown menu                 | [Item[]](../../composables/useDropdownItems)     | -      | () => []  |
| menuClass | The classes that will be forwarded to the menu | `string`\|`array`\|`object`                      | -      | undefined |
| size      | The size of the component                      | [Size](../../composables/useSize) (string)       | -      | 'md'      |
| variant   | The variant of the component                   | [Variant](../../composables/useVariant) (string) | -      | 'primary' |

<!-- TODO Auto-generate based on type? -->

## Events

| Name   | Args | Description                                                                                        |
| ------ | ---- | -------------------------------------------------------------------------------------------------- |
| show   |      | Fires immediately when the `show` instance method is called                                        |
| shown  |      | Fired when the dropdown has been made visible to the user and CSS transitions have completed       |
| hide   |      | Fires immediately when the `hide` instance method has been called                                  |
| hidden |      | Fired when the dropdown has finished being hidden from the user and CSS transitions have completed |

## Slots

| Name           | Description                                 | Bindings                                 |
| -------------- | ------------------------------------------- | ---------------------------------------- |
| dropdownToggle | Displays the dropdown toggle                |                                          |
| default        | Displays content inside the dropdown toggle |                                          |
| dropdownMenu   | Displays the dropdown menu                  | <br/>**items** `array` - The `Item` list |
| prepend        | _Forwarded to child_                        |                                          |
| items          | _Forwarded to child_                        |                                          |
| item           | _Forwarded to child_                        |                                          |
| append         | _Forwarded to child_                        |                                          |
