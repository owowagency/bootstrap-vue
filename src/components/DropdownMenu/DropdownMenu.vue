<template>
    <ul class="dropdown-menu">
        <!-- @slot Prepends the dropdown menu -->
        <slot name="prepend" />

        <!--
        @slot Displays content inside the dropdown menu
        @binding {array} items The `Item` list
        @binding {function} click-item The function to call when an item is clicked
        -->
        <slot
            name="items"
            :click-item="clickItem"
            :items="items"
        >
            <template
                v-for="(item, index) of items"
                :key="`item-${index.toString()}`"
            >
                <!--
                @slot Displays the dropdown menu item
                @binding {object} item The `Item` object
                @binding {function} click-item The function to call when an item is clicked
                -->
                <slot
                    name="item"
                    :click-item="clickItem"
                    :item="item"
                >
                    <DropdownMenuItem
                        v-bind="item"
                        :item-class="itemClass"
                        @click="clickItem(item)"
                    />
                </slot>
            </template>
        </slot>

        <!-- @slot Appends the dropdown menu -->
        <slot name="append" />
    </ul>
</template>

<script lang="ts">
import {Item, dropdownItemsProps} from '@/composables/useDropdownItems';
import DropdownMenuItem from '@/components/DropdownMenuItem';
import {dropdownItemProps} from '@/composables/useDropdownItem';
</script>

<script lang="ts" setup>
defineProps({
    ...dropdownItemProps,
    ...dropdownItemsProps,
});

const emit = defineEmits([
    /**
     * Fired when an item is clicked
     * @param {Item} item The item that was clicked
     */
    'click:item',
]);

const clickItem = (item: Item) => emit('click:item', item);
</script>

<docs>
```vue
<DropdownMenu :items="[Item]" />
```

## Imported Props

| Prop name | Description                    | Type                                         | Values | Default  |
| --------- | ------------------------------ | -------------------------------------------- | ------ | -------- |
| items     | The items of the dropdown menu | [Item[]](../../composables/useDropdownItems) | -      | () => [] |
</docs>
