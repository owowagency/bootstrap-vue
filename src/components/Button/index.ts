import Button from './Button.vue';

export * from './Button.vue';
export default Button;
