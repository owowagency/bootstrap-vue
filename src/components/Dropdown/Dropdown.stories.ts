import Dropdown from '.';
import DropdownMenu from '@/components/DropdownMenu';
import {sizes} from '@/composables/useSize';
import {variants} from '@/composables/useVariant';

export default {
    title: 'Components/Dropdown',
    component: Dropdown,
    argTypes: {},
};

const defaultArgs = {
    items: [
        {label: 'Item 1'},
        {label: 'Item 2'},
        {label: 'Item 3'},
    ],
    text: 'I am a dropdown',
};

const template = (args) => ({
    components: {Dropdown},
    setup: () => ({args}),
    template: '<Dropdown v-bind="args" />',
});

export const Default = template.bind({});

Default.args = defaultArgs;

export const Variants = (args) => ({
    components: {Dropdown},
    setup: () => ({args, variants}),
    template: '<Dropdown v-for="v of variants" class="me-1 mb-1 d-inline-block" :variant="v" v-bind="args" />',
});

Variants.args = defaultArgs;

export const Sizing = (args) => ({
    components: {Dropdown},
    setup: () => ({args, sizes}),
    template: '<Dropdown v-for="s of sizes" class="me-1 mb-1 d-inline-block" :size="s" v-bind="args" />',
});

Sizing.args = defaultArgs;

export const Menu = (args) => ({
    components: {DropdownMenu},
    setup: () => ({args}),
    template: '<DropdownMenu class="d-block" v-bind="args" />',
});

Menu.args = {
    items: [
        {label: 'Dropdown header', header: true},
        {label: 'Item one'},
        {label: 'Item two'},
        {label: 'Active item', active: true},
        {label: 'Item three'},
        {divider: true},
        {label: 'Item four'},
        {label: 'Item five'},
    ],
};
