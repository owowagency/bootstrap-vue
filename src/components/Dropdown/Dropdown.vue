<template>
    <div
        ref="dropdown"
        class="dropdown"
    >
        <!-- @slot Displays the dropdown toggle -->
        <slot name="dropdownToggle">
            <Btn
                class="dropdown-toggle"
                :class="toggleClass"
                data-bs-toggle="dropdown"
                :size="size"
                :variant="variant"
            >
                <!-- @slot Displays content inside the dropdown toggle -->
                <slot>
                    Select
                </slot>
            </Btn>
        </slot>

        <!--
        @slot Displays the dropdown menu
        @binding {array} items The `Item` list
        -->
        <slot
            name="dropdownMenu"
            v-bind="{'onClick:item': $attrs['onClick:item']}"
            :items="items"
        >
            <DropdownMenu
                :items="items"
                :class="menuClass"
                :item-class="itemClass"
                v-bind="{'onClick:item': $attrs['onClick:item']}"
            >
                <!-- Rollup does not dynamically overriding child slots. See issue #35
                <template
                    v-for="slotName in menuSlots"
                    #[slotName]="slotScope"
                >
                    <slot
                        :name="slotName"
                        v-bind="slotScope"
                    />
                </template> -->

                <template #prepend="slotScope">
                    <!-- @slot *Forwarded to child* -->
                    <slot
                        name="prepend"
                        v-bind="slotScope"
                    />
                </template>

                <template #items="slotScope">
                    <!-- @slot *Forwarded to child* -->
                    <slot
                        name="items"
                        v-bind="slotScope"
                    />
                </template>

                <template #item="slotScope">
                    <!-- @slot *Forwarded to child* -->
                    <slot
                        name="item"
                        v-bind="slotScope"
                    />
                </template>

                <template #append="slotScope">
                    <!-- @slot *Forwarded to child* -->
                    <slot
                        name="append"
                        v-bind="slotScope"
                    />
                </template>
            </DropdownMenu>
        </slot>
    </div>
</template>

<script lang="ts">
import Btn from '@/components/Button';
import DropdownMenu from '@/components/DropdownMenu';
import {dropdownItemProps} from '@/composables/useDropdownItem';
import {dropdownProps} from '@/composables/useDropdown';
import {ref} from 'vue';
import useBootstrapEmits from '@/composables/useBootstrapEmits';
import useBootstrapInstance from '@/composables/useBootstrapInstance';

const dropdownEvents = ['show', 'shown', 'hide', 'hidden'] as const;

type DropdownEvent = typeof dropdownEvents[number];
</script>

<script lang="ts" setup>

defineProps({
    ...dropdownItemProps,
    ...dropdownProps,
});

const emit = defineEmits<{(event: DropdownEvent): void}>();

const dropdown = ref<HTMLElement>();

useBootstrapEmits(
    dropdown,
    dropdownEvents,
    emit,
    'dropdown',
);

const {bsInstance: bsDropdown} = useBootstrapInstance(
    'Dropdown',
    dropdown,
);

// Rollup does not like dynamically overriding slots so this is not used for now.
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const menuSlots = [
    'prepend',
    'items',
    'item',
    'append',
];

defineExpose({bsDropdown});
</script>

<docs>
```vue
<Dropdown :items="[Item]" />
```

## Imported Props

| Prop name | Description                                    | Type                                             | Values | Default   |
| --------- | ---------------------------------------------- | ------------------------------------------------ | ------ | --------- |
| items     | The items of the dropdown menu                 | [Item[]](../../composables/useDropdownItems)     | -      | () => []  |
| menuClass | The classes that will be forwarded to the menu | `string`\|`array`\|`object`                      | -      | undefined |
| size      | The size of the component                      | [Size](../../composables/useSize) (string)       | -      | 'md'      |
| variant   | The variant of the component                   | [Variant](../../composables/useVariant) (string) | -      | 'primary' |

<!-- TODO Auto-generate based on type? -->
## Events

| Name   | Args | Description                                                                                        |
| ------ | ---- | -------------------------------------------------------------------------------------------------- |
| show   |      | Fires immediately when the `show` instance method is called                                        |
| shown  |      | Fired when the dropdown has been made visible to the user and CSS transitions have completed       |
| hide   |      | Fires immediately when the `hide` instance method has been called                                  |
| hidden |      | Fired when the dropdown has finished being hidden from the user and CSS transitions have completed |
</docs>
