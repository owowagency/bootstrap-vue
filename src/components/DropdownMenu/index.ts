import DropdownMenu from './DropdownMenu.vue';

export default DropdownMenu;
