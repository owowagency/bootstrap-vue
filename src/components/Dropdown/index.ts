import Dropdown from './Dropdown.vue';

export default Dropdown;
