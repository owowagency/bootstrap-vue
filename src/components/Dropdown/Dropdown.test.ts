import {describe, expect, it, vi} from 'vitest';
import Button from '@/components/Button';
import Dropdown from '.';
import DropdownMenu from '@/components/DropdownMenu';
import {shallowMount} from '@vue/test-utils';

vi.mock('bootstrap', () => ({
    Dropdown: {
        getOrCreateInstance: vi.fn(),
    },
    __v_isShallow: true,
}));

const items = [
    {label: 'Item 1'},
    {label: 'Item 2'},
    {label: 'Item 3'},
];

describe('template', () => {
    componentRenderTest(Dropdown, {props: {items}});

    componentSlotRenderTest(Dropdown);

    componentSlotRenderTest(Dropdown, 'dropdownMenu', {shallow: false});

    componentSlotRenderTest(Dropdown, 'prepend', {shallow: false});

    componentSlotRenderTest(Dropdown, 'items', {shallow: false});

    componentSlotRenderTest(Dropdown, 'item', {
        props: {items},
        shallow: false,
    });

    componentSlotRenderTest(Dropdown, 'append', {shallow: false});

    ['show', 'shown', 'hide', 'hidden'].forEach((event: string) => {
        componentBootstrapEventTest(
            Dropdown,
            '.dropdown',
            event,
            'dropdown',
        );
    });

    it('adds menu class to DropdownMenu', () => {
        const wrapper = shallowMount(Dropdown, {
            props: {
                menuClass: 'hello-there',
                items,
            },
        });

        const dropdownMenu = wrapper.findComponent(DropdownMenu);

        expect(dropdownMenu.attributes('class')).toBe('hello-there');
    });

    it('adds toggle class to toggle', () => {
        const wrapper = shallowMount(Dropdown, {
            props: {
                toggleClass: 'hello-there',
                items,
            },
        });

        const dropdownMenu = wrapper.findComponent(Button);

        expect(dropdownMenu.classes()).toContain('hello-there');
    });

    it('forwards item class to DropdownMenu', () => {
        const wrapper = shallowMount(Dropdown, {
            props: {
                itemClass: 'hello-there',
                items,
            },
        });

        const dropdownMenu = wrapper.findComponent(DropdownMenu);

        expect(dropdownMenu.vm.itemClass).toBe('hello-there');
    });
});

describe('onMounted', () => {
    it('sets bsDropdown', async() => {
        const wrapper = await shallowMount(Dropdown);

        expect((await import('bootstrap')).Dropdown.getOrCreateInstance).toBeCalledWith(wrapper.vm.$refs.dropdown, {});

        // TODO: Unable to assert bsDropdown value.
    });
});
