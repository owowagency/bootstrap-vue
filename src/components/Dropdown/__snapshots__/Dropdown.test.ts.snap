// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`template > renders default 1`] = `
"<div class=\\"dropdown\\">
  <!-- @slot Displays the dropdown toggle -->
  <button-stub disabled=\\"false\\" outline=\\"false\\" size=\\"md\\" variant=\\"primary\\" class=\\"dropdown-toggle\\" data-bs-toggle=\\"dropdown\\"></button-stub>
  <!--
        @slot Displays the dropdown menu
        @binding {array} items The \`Item\` list
        -->
  <dropdown-menu-stub items=\\"[object Object],[object Object],[object Object]\\"></dropdown-menu-stub>
</div>"
`;
